### ISSUE TITLE
[consistent-type-assertions] Feature Request: rule should have an option to get "assertionStyle": "never" behaviour, but allow casting from unknown.

### ISSUE THREAD

**Issue by pcone** [typescript-eslint/typescript-eslint#4688]
- [N/A] I have tried restarting my IDE and the issue persists.
- [✔️] I have updated to the latest version of the packages.
- [✔️] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

I have disallowed casting in a codebase with the consistent-type-assertions rule. This is useful in 95% of code because it's usually better to type things correctly so that the casting isn't needed, and the casting can hide bugs when the value you're operating on doesn't match the type that you're assuming it can be.

However, when you have a value of type 'unknown' the *only* thing you can do with it is cast (or check via instanceOf, but that only works in codebases that consistently use Classes with constructors rather than raw objects).

It would be nice if I could allow casting without a suppression in these cases - mostly catch blocks and type guards. 
I know I can just explicitly type the inputs as any instead, but then code can be written where it isn't obvious that fields that might not exist are being accessed, as opposed to explicitly casting on just the line where it's needed. (I believe similar thoughts triggered the switch from any to unknown in catch blocks https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#use-unknown-catch-variables).

Current:
```JSON
{
  "rules": {
    "@typescript-eslint/consistent-type-assertions": [
            "error",
            {
                "assertionStyle": "never"
            }
    ]
  }
}
```

Proposed:
```JSON
{
  "rules": {
    "@typescript-eslint/consistent-type-assertions": [
            "error",
            {
                "assertionStyle": "onlyFromUnknown"
            }
    ]
  }
}
```

Catch blocks like this would be allowed with proposed option:
```TS
try {
  //do something
} catch(e /*unknown instead of any by default as of TS 4.4*/) {
  if(typeof (e as Foo).message === "string") {
    //do some stuff with the message (contrived example but you get the point)
  }
}
```

Likewise, type guards like the following would also be allowed without a suppression:
```TS
function isFoo(value: unknown): value is Foo {
  return typeof (value as Foo).uniqueField !== "undefined";
}
```

All casts from anything other than unknown would still trigger an error:
```TS
function doFoo(value: Foo) {
  const fooAsBar = Foo as Bar
}

**Comment by bradzacher** [comment]
> However, when you have a value of type 'unknown' the only thing you can do with it is cast (or check via instanceOf ...

This isn't correct!

`typeof` will let you narrow the type to the primitive types:
```ts
declare const x: unknown;

if (typeof x === 'string') {
  // x is now a string
}
if (typeof x === 'boolean') {
  // x is now a boolean
}
if (typeof x === 'number') {
  // x is now a number
}
```

Additionally you could define [type guards (aka predicate functions)](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards):
```ts
interface Foo {
  prop: string;
}

function isFoo(x: any): x is Foo {
  return (
    typeof x === 'object' &&
    x !== null &&
    Object.hasOwn(x, 'prop') &&
    typeof x.prop === 'string'
  );
}

declare const x: unknown;

if (isFoo(x)) {
  // x is of type Foo now
}
```

----

Making this change would require adding type-information to the rule, which we can't do (breaking change amongst other things) - so we can't make this change.

Thanks for the issue!
